sinon/lib/base.py

```diff
@@ -75,6 +75,22 @@ def __new__(cls, obj=None, prop=None, func=None):
         cls._queue.append(new)
         return weakref.proxy(new)
 
+    def __get_wrapper(self):
```

Comment:
From OOD view, I will suggest the following change:

Changing the function name from **private**: `__get_wrapper(self)` to **protected** - `_get_wrapper(self)`, since it will be used by spy or other component in the lib.

And later, changing all `_SinonBase__get_wrapper()` to `_get_wrapper()`, since it has become protected function.

Reply:
Thanks, Kir, I just learned something new about Python. I find invoking the `super` constructor a bit strange, but if that's the standard idiom then ok.

Reply:
Thank you too. Your changes make me go back to review lots concept about python.

I am not sure about the word `strange` you mean here.

Here is the concept of OOP:
```python
class Parent(object):
    def say(self):
        self.word = "Hello"

class Child(Parent):
    def say(self):
        self.word = "Hello World!!"

parent = Parent()
parent.say()
child = Child()
child.say()
```
* Why child can call `say` directly?
Because child defines its own function with same name in parent, but there is no relation between both.

---

However, the case might be different if there is some relations between Parent and Child:
```python
class Parent(object):
    def __init__(self):
        self.say_cnt = 0
    def say(self):
        self.word = "Hello"
        self.say_cnt += 1

class Child(Parent):
    def say(self):
        self.word = "Hello World!!"

parent = Parent()
parent.say()
print(parent.say_cnt) #1

child = Child()
child.say()
print(child.say_cnt) #0
```

---

In this case, Child will not call parent's `say`, but for achieving normal action on `say_cnt` in Parent, we can use `super(Child, self).say()` in Child.

```python
class Child(Parent):
    def say(self):
        super(Child, self).say()
        self.word = "Hello World!!"
```

Reply:
Hi Kir,

This is just my opinion, and it is not very important.

There are a lot of things I find weird about Python's object-oriented model, and the scenario you just showed is one of them! :) Python is supposed to be a language that prides itself on being simple and succinct. Yet, Python forces us to declare the `self` parameter all the time in every method declaration, for no reason. When calling the super class, why require us to write `super(Child, self)` instead of just `super()`?

Java is an extremely verbose language, where you have to type in `System.out.println` just to print something to the console. Yet, in Java I can just call `super()`. In Java, all of a parent class's methods are automatically copied over to its children, so that I could just write `this.parentMethod()` instead of `super(Child, self).parent_method()`.

I find it weird, that's all. :)

--Jonathan

Reply:
> When calling the super class, why require us to write super(Child, self) instead of just super()?

That's for compatibility reason only, in python3, you can write `super()` only. (https://stackoverflow.com/questions/10482953/python-extending-with-using-super-python-3-vs-python-2)

My code contains a lot of this idea, because I need to consider python2's compatibility.

> In Java, all of a parent class's methods are automatically copied over to its children

This idea is as same as python. 

I believe It's an opinion-based thing. I prefer to write code in an explicit way. 
* If you are using parent function, we should tell others in the code. 
* If you are using class variable, we should tell others in the code.
* If the function is a staticmethod, we should not implement them as an instance method.
* If the function is a classmethod, we should not implement them as an instance method.
... (there are several examples)

**Python is a flexible language. Thus we have no need to follow any of them in above.** However, in my experience, while people trace code, writing them in an explicit way will enhance the readability. That's why I suggest you to modify in that way.

In modern python discussion, lots people are trying to promote [type hint](https://www.python.org/dev/peps/pep-0484/#type-definition-syntax) in conference. The reason is same, the language is flexible, but it can be solid too.

Reply:
Hi Kir, I submitted an update for the Pull Request, so it is ready for review.